src/app/clusters/general-diagnostics-server/general-diagnostics-server.cpp

```diff
@@ -385,9 +392,42 @@ bool emberAfGeneralDiagnosticsClusterTestEventTriggerCallback(CommandHandler * c
 bool emberAfGeneralDiagnosticsClusterTimeSnapshotCallback(CommandHandler * commandObj, ConcreteCommandPath const & commandPath,
                                                           Commands::TimeSnapshot::DecodableType const & commandData)
 {
-    // TODO(#30096): Command needs to be implemented.
-    ChipLogError(Zcl, "TimeSnapshot not yet supported!");
-    commandObj->AddStatus(commandPath, Status::InvalidCommand);
+    ChipLogError(Zcl, "Received TimeSnapshot command!");
+
+    Commands::TimeSnapshotResponse::Type response;
+
+    System::Clock::Microseconds64 posix_time_us{ 0 };
+
+#ifdef TIME_SYNCHRONIZATION_CLUSTER_INCLUDED
+#ifdef ZCL_USING_TIME_SYNCHRONIZATION_CLUSTER_SERVER
+    bool time_is_synced = false;
+    using Clusters::TimeSynchronization::GranularityEnum;
+    GranularityEnum granularity = Clusters::TimeSynchronization::TimeSynchronizationServer::Instance().GetGranularity();
+    time_is_synced              = (granularity == GranularityEnum::kSecondsGranularity) ||
+        (granularity == GranularityEnum::kMillisecondsGranularity) || (granularity == GranularityEnum::kMicrosecondsGranularity);
```

Comment:
This might be better as a switch statement, so if a new granularity value is added the compiler will catch the missing case...

Reply:
Done.